examples/nanoChatGPT/env.py

```diff
@@ -46,20 +48,15 @@ def _step(self, tensordict):
 @torch.no_grad()
 def _reset(self, tensordict):
     self.step_num = 0
-    if tensordict is None or tensordict.is_empty():
-        # if no tensordict is passed, we generate a single set of hyperparameters
-        # Otherwise, we assume that the input tensordict contains all the relevant
-        # parameters to get started.
-        tensordict = TensorDict({}, batch_size=self.config["batch_size"])
-
     batch = next(self.dataloader)
 
+    prompt = batch.prompt[:self.batch_size[0], -self.block_size :]
     out = TensorDict(
         {
-            "prompt": batch.prompt[:, -self.config["block_size"] :],
-            "done": torch.zeros((*batch.prompt.shape[:-1], 1, 1), dtype=torch.bool),
```

Comment:
why shape (1, 1) at the end?

Reply:
we had issues with shapes

Reply:
It's better if we stay consistent IMO

Here is the recap of the workflow I understood:


Our "state" (the prompt) has shape [B, Tokens]
It is embedded in [B, Tokens, F] and then passed through the transformer which returns another [B, Tokens, F] tensor

- For the rewards, we take the last token ([:, -1, :]) of this final tensor [B, F] and get a single reward through a reward layer, hence we end up with [B, 1] shape
- For the value it's pretty much the same story.
- For the action, we have the [B, Tokens, F] tensor. We take the last token, and get the logits of shape [B, 1, L] (note that here we index with [-1] whereas for rewards and value we index with -1: I would suggest using -1 here too). Then out of this [B, 1, L] we sample an action which has size [B, 1], and we get a log_prob of size [B, 1] too. If we were using the -1 index these would be [B] and I personally would feel more comfortable with that.
- So our batch is 
```
TensorDict({
  "action": Tensor([B, 1]),
  "log_prob": Tensor([B, 1]),
  "reward": Tensor([B, 1]),
  "done": Tensor([B, 1]),
  "prompt": Tensor([B, Tokens]),
}, [B])
```
Add a `T` dimension after `B` for a rollout.

I believe that the reason why you were unsqueezing the reward and done was to get [B, 1, 1] like we eventually have for log_prob, but that's one level of unsqueezing too much.

The problem I was mentioning before regarding PPO for which I added a custom patch was that in PPO we unsqueeze the log_prob, but as I mentioned above I think this problem will vanish if we index the input to lm_head with -1 intead of [-1]: our log_prob will have size [B] and we'll be able to pass it to the PPO loss without trouble.


TLDR: let's try to change this line:
```
            logits = self.lm_head(x[:, [-1], :]) # note: using list [-1] to preserve the time dim
```
to
```
            logits = self.lm_head(x[:, -1, :]) # note: using list [-1] to preserve the time dim
```
and see what happens